### PR TITLE
jupyter-docker: break image into 'base' and 'final'

### PR DESCRIPTION
**Pull request by eap**
Many application-specific packages have been added to the Jupyter base-image, but I think that builders of Leo would benefit from separation between "_stuff needed to run a Leonardo notebook_" and "_stuff that analysts want_".

This change starts the process of carving these groups apart. The base image still contains many analyst-centered package installs, but now there's a general framework for analyst tool installation that is decoupled from notebook server config.

In all cases:
- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: jupyter-docker/Dockerfile.final [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+USER $USER
+
+##########################
+# Python and R extensions
+##########################
+
+RUN pip2 install --user \
+        google-cloud \
```
Comment (rtitle) [inline jupyter-docker/Dockerfile.final]: I think the base image already installs google-cloud?

https://github.com/DataBiosphere/leonardo/blob/develop/jupyter-docker/Dockerfile#L122
https://github.com/DataBiosphere/leonardo/blob/develop/jupyter-docker/Dockerfile#L151

**Comment by rtitle** [comment]
This is indeed the direction we want to move in. With this framework I could see us moving other things from `Dockerfile.base` to `Dockerfile.final`. In the future we could even let the user pick the final image from a few options, or specify their own.

So yeah, I'm happy to merge this -- will just kick off our integration tests to make sure there are no issues.

File: jupyter-docker/Dockerfile.final (hunk 1 above)
Comment (eap) [reply jupyter-docker/Dockerfile.final]: Oh, very true. I'll push a commit to remove this line.

**Comment by rtitle** [comment]
I ran a build with this change and got an error:
```
Step 1 : ARG BASE_IMAGE
Please provide a source image with `from` prior to commit
Build step 'Execute shell' marked build as failure
```
https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-build/2008/console

When was `ARG` added to docker? I think we're using a rather old version (1.10), but are in the process of ugrading it.

**Comment by eap** [comment]
The parameter option has been in since at least 2017.

**Comment by eap** [comment]
jenkins retest

**Comment by rtitle** [comment]
I am closing this PR for now, but referencing it in https://github.com/DataBiosphere/leonardo/issues/414 which is to split up our Dockerfile and support user-specified images.

**Pull request closed without merging**